framework/src/modules/interoperability/mainchain/module.ts

```diff
@@ -227,4 +243,174 @@ export class MainchainInteroperabilityModule extends BaseInteroperabilityModule
 			],
 		};
 	}
+
+	// @see https://github.com/LiskHQ/lips/blob/main/proposals/lip-0045.md#mainchain
+	// eslint-disable-next-line @typescript-eslint/require-await
+	public async initGenesisState(ctx: GenesisBlockExecuteContext): Promise<void> {
+		const genesisBlockAssetBytes = ctx.assets.getAsset(MODULE_NAME_INTEROPERABILITY);
+		if (!genesisBlockAssetBytes) {
+			return;
+		}
+
+		const genesisInteroperability = codec.decode<GenesisInteroperability>(
+			genesisInteroperabilitySchema,
+			genesisBlockAssetBytes,
+		);
+
+		validator.validate<GenesisInteroperability>(
+			genesisInteroperabilitySchema,
+			genesisInteroperability,
+		);
+
+		const { ownChainName, ownChainNonce, chainInfos } = genesisInteroperability;
+
+		// On the mainchain, the following checks are performed:
+		if (ctx.chainID.equals(getMainchainID(ctx.chainID))) {
+			// ownChainName == CHAIN_NAME_MAINCHAIN.
+			if (ownChainName !== CHAIN_NAME_MAINCHAIN) {
+				throw new Error(`ownChainName must be equal to CHAIN_NAME_MAINCHAIN'`);
+			}
+
+			// if chainInfos is empty, then ownChainNonce == 0
+			if (chainInfos.length === 0) {
+				if (ownChainNonce !== BigInt(0)) {
+					throw new Error(`ownChainNonce must be 0 if chainInfos is empty.`);
+				}
+			}
+
+			// If chainInfos is non-empty
+			// ownChainNonce > 0
+			if (ownChainNonce <= 0) {
+				throw new Error(`ownChainNonce must be positive if chainInfos is not empty.`);
+			}
+
+			// Each entry chainInfo in chainInfos has a unique chainInfo.chainID
+			const chainIDs = chainInfos.map(info => info.chainID);
+			if (!bufferArrayUniqueItems(chainIDs)) {
+				throw new Error(`chainInfos doesn't hold unique chainID.`);
+			}
+
+			// chainInfos should be ordered lexicographically by chainInfo.chainID
+			const sortedByChainID = [...chainInfos].sort((a, b) => a.chainID.compare(b.chainID));
+			for (let i = 0; i < chainInfos.length; i += 1) {
+				if (!chainInfos[i].chainID.equals(sortedByChainID[i].chainID)) {
+					throw new Error('chainInfos is not ordered lexicographically by chainID.');
+				}
+			}
+
```

Comment:
Add check that among all items of chainInfos array, the chainData.name properties are pairwise distinct.  It was removed from `verifyChainData()` but not added here.